### ISSUE TITLE
Build and tests are failing

### ISSUE THREAD

**Issue by gotthardp** [tpm2-software/tpm2-openssl#5]
Due to recent changes in OpenSSL the TPM2 provider cannot be built and tests fail with the latest master.

As a workaround, please use the OpenSSL from 6.6.2020, e.g. https://github.com/openssl/openssl/commit/97cf9b05fa1cdb8e4e7f60016aa95ae0e976e8c3. That version worked.

**Comment by levitte** [comment]
Hmmmm, there are some things that don't quite fit with the model OpenSSL runs with...  I think.
Especially considering this, it seems like what you have is a separate key type, at least as far as OpenSSL is concerned:

https://github.com/tpm2-software/tpm2-openssl/blob/618b27c8e8dd595411a1a1c192a1360004b41138/src/tpm2-provider-pkey.c#L28

... but also considering the TSS2 PEM name.

So I think that the appropriate way out of your troubles is to adopt that thought, and fully go for a TPM2 key type instead of trying to obfuscate RSA, EC, yada yada.  That gives you quite a lot of power to decide on your own exactly what can be exported/imported through KEYMGMT, and adapt the decoders and encoders accordingly.

**Comment by levitte** [comment]
I wouldn't mind spending some time helping out with this, as this is also a good way to test aspects of OpenSSL's STORE and DECODER code that we may have missed.

**Comment by levitte** [comment]
Side note: finding this in the DER decoder was a bit of a surprise:

https://github.com/tpm2-software/tpm2-openssl/blob/618b27c8e8dd595411a1a1c192a1360004b41138/src/tpm2-provider-decoder-der.c#L68

**Comment by gotthardp** [comment]
@levitte Thank you. Any help would be greatly appreciated. I can do the coding if you show me the right way as it's not always obvious.

The ultimate goal is that the TPM2-based keys work as standard RSA or EC keys in TLS or X.509 operations. Since it *is* a RSA key (just a different implementation) I thought it should be named RSA.

**Comment by levitte** [comment]
I identified one problem in OpenSSL itself...  encoders and decoders couldn't have multiple names, the code to allow that had simply not been put in place.  Because of that, this decoder never registered:

https://github.com/tpm2-software/tpm2-openssl/blob/618b27c8e8dd595411a1a1c192a1360004b41138/src/tpm2-provider.c#L229

openssl/openssl#15904 solves that problem.

**Comment by levitte** [comment]
Also, I'm currently working with the assumption that your 'file' store implementation is unnecessary, so I've patched like this:

``` diff
diff --git a/src/tpm2-provider.c b/src/tpm2-provider.c
index 311aa29..db87a37 100644
@@ -234,7 +234,9 @@ extern const OSSL_DISPATCH tpm2_file_store_functions[];
 extern const OSSL_DISPATCH tpm2_object_store_functions[];
 
 static const OSSL_ALGORITHM tpm2_stores[] = {
+#if 0
     { "file", TPM2_PROPS(store), tpm2_file_store_functions },
+#endif
     { "object", TPM2_PROPS(store), tpm2_object_store_functions },
     { "handle", TPM2_PROPS(store), tpm2_object_store_functions },
     { NULL, NULL, NULL }
```

**Comment by levitte** [comment]
Furthermore, I found another bug in OpenSSL, which affects this test line:

https://github.com/tpm2-software/tpm2-openssl/blob/618b27c8e8dd595411a1a1c192a1360004b41138/test/rsa_genrsa_check.sh#L60

It's possible to reproduce it without TPM2, I've raised an issue on that in openssl/openssl#15907

I'm on holyday until Monday, so unless someone else picks it up, don't expect to see a fix before Monday

**Comment by levitte** [comment]
All this being said, with openssl/openssl#15904, the patch taking out the 'file' store implementation, #6 and `-provider base` added in quite a log of places in the test scripts (PR upcoming), I got quite a bit of the way, half of the tests passing!

``` console:
$ LD_LIBRARY_PATH=$HOME/.private-install/opt/openssl-3.0/bin PATH=$HOME/.private-install/opt/openssl-3.0/bin:$PATH OPENSSL_MODULES=./.libs make check
make  test/selftest
make[1]: Går till katalogen ”/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl”
make[1]: ”test/selftest” är aktuell.
make[1]: Lämnar katalogen ”/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl”
make  check-TESTS
make[1]: Går till katalogen ”/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl”
make[2]: Går till katalogen ”/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl”
PASS: test/list.sh
PASS: test/rand.sh
FAIL: test/rsa_genrsa_check.sh
PASS: test/rsa_genpkey_sign.sh
PASS: test/rsa_genpkey_sign_rawin.sh
FAIL: test/rsa_genpkey_auth_parent.sh
FAIL: test/rsa_createak_auth.sh
FAIL: test/rsa_createak_sign_object.sh
FAIL: test/rsa_createak_sign_handle.sh
FAIL: test/rsa_create_decrypt.sh
PASS: test/rsa_genpkey_x509_cert.sh
FAIL: test/rsa_genpkey_x509_cmp.sh
FAIL: test/rsa_genpkey_tls_server.sh
FAIL: test/rsa_createak_x509_csr.sh
PASS: test/rsapss_genpkey_sign_rawin.sh
FAIL: test/rsapss_createak_tls_server.sh
FAIL: test/rsa_pki/rsa_pki.sh
PASS: test/ec_genpkey_check.sh
PASS: test/ec_genpkey_parameters.sh
PASS: test/ecdsa_genpkey_auth.sh
PASS: test/ecdsa_genpkey_sign_rawin.sh
FAIL: test/ecdsa_createak_sign_handle.sh
PASS: test/ecdh_genpkey_keyexch.sh
FAIL: test/ec_createak_x509_csr.sh
FAIL: test/ec_genpkey_tls_server.sh
FAIL: test/ec_pki/ec_pki.sh
PASS: test/store_errors.sh
PASS: test/digest.sh
PASS: test/cipher_aes128_ecb.sh
PASS: test/cipher_aes128_cbc.sh
PASS: test/cipher_aes256.sh
PASS: test/cipher_aes256_nopad.sh
FAIL: test/cipher_camellia128.sh
FAIL: test/selftest
============================================================================
Testsuite summary for tpm2-openssl 2.0.0
============================================================================
# TOTAL: 34
# PASS:  17
# SKIP:  0
# XFAIL: 0
# FAIL:  17
# XPASS: 0
# ERROR: 0
============================================================================
See ./test-suite.log
Please report to https://github.com/gotthardp/tpm2-openssl/issues
============================================================================
make[2]: *** [Makefile:1121: test-suite.log] Fel 1
make[2]: Lämnar katalogen ”/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl”
make[1]: *** [Makefile:1229: check-TESTS] Fel 2
make[1]: Lämnar katalogen ”/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl”
make: *** [Makefile:1442: check-am] Fel 2
```

Going through the logs, it seems that some of the failures are outside my control...  I see a number of these:

```
ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
```

A fuller example, this taken from `test/rsa_create_decrypt.log`:

```
++ tpm2_evictcontrol -c testkey.ctx
++ cut -d ' ' -f 2
++ head -n 1
WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
ERROR: Unable to run tpm2_evictcontrol
```

... and considering that it's the log from this lines:

https://github.com/tpm2-software/tpm2-openssl/blob/618b27c8e8dd595411a1a1c192a1360004b41138/test/rsa_create_decrypt.sh#L17

... it's no wonder that the rest of the same log shows failure (because no HANDLE):

```
+ HANDLE=
+ openssl pkey -provider tpm2 -in handle: -pubout -out testkey.pub
WARNING:esys:src/tss2-esys/api/Esys_ReadPublic.c:320:Esys_ReadPublic_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/esys_tr.c:231:Esys_TR_FromTPMPublic_Finish() Error ReadPublic ErrorCode (0x00000184) 
ERROR:esys:src/tss2-esys/esys_tr.c:321:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x00000184) 
Could not read key from handle:
80D21790FC7E0000:error:4180000C:tpm2::cannot load key::-1:388 tpm:handle(1):value is out of range or is not correct for the context
FAIL test/rsa_create_decrypt.sh (exit status: 1)
```

**Comment by levitte** [comment]
Hah, another tid-bit had me find another small OpenSSL bug:

```
Starting /home/levitte/gitwrk/github.com/tpm2-software/tpm2-openssl/test/rsa_genpkey_x509_cmp.sh
+ openssl req -provider tpm2 -provider base -x509 -newkey rsa:2048 -sha256 -nodes -subj /C=GB/CN=root.example.com -extensions v3_ca -keyout test-ca-key.pem -out test-ca-cert.pem
crypto/evp/evp_fetch.c:110: OpenSSL internal error: Assertion failed: name_id > 0 && name_id <= METHOD_ID_NAME_MAX
/home/levitte/gitwrk/github.com/tpm2-software/tpm2-openssl/test/rsa_genpkey_x509_cmp.sh: rad 8: 2786449 Avbruten (SIGABRT)      openssl req -provider tpm2 -provider base -x509 -newkey rsa:2048 -sha256 -nodes -subj "/C=GB/CN=root.example.com" -extensions v3_ca -keyout test-ca-key.pem -out test-ca-cert.pem
FAIL test/rsa_genpkey_x509_cmp.sh (exit status: 134)
```

Fix in openssl/openssl#15912

**Comment by levitte** [comment]
So far, I have the following fixes:

- openssl/openssl#15904
- openssl/openssl#15912
- #6 
- #7
- #8

**Comment by gotthardp** [comment]
Thank you @levitte!

Using `-provider base` feels easy enough to me so that we could get rid of the file store completely. So far I used `provider default -propquery ?provider=tpm2` which is terribly long for users to type on a command line.

**Comment by levitte** [comment]
Yeah.  I'll go through those places and see how easy they are modified, it's possible that there are cases where `-provider default` is necessary (but not `-propquery ?provider=tpm2`).

That won't happen now, though.  I'm off for the rest of the day.

**Comment by levitte** [comment]
With all the current fixes combined, I'm now at this:

``` console
$ LANG=C LD_LIBRARY_PATH=$HOME/.private-install/opt/openssl-3.0/bin PATH=$HOME/.private-install/opt/openssl-3.0/bin:$PATH OPENSSL_MODULES=./.libs make check
make  test/selftest
make[1]: Entering directory '/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl'
make[1]: 'test/selftest' is up to date.
make[1]: Leaving directory '/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl'
make  check-TESTS
make[1]: Entering directory '/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl'
make[2]: Entering directory '/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl'
PASS: test/list.sh
PASS: test/rand.sh
FAIL: test/rsa_genrsa_check.sh
PASS: test/rsa_genpkey_sign.sh
PASS: test/rsa_genpkey_sign_rawin.sh
FAIL: test/rsa_genpkey_auth_parent.sh
FAIL: test/rsa_createak_auth.sh
FAIL: test/rsa_createak_sign_object.sh
FAIL: test/rsa_createak_sign_handle.sh
FAIL: test/rsa_create_decrypt.sh
PASS: test/rsa_genpkey_x509_cert.sh
FAIL: test/rsa_genpkey_x509_cmp.sh
PASS: test/rsa_genpkey_tls_server.sh
FAIL: test/rsa_createak_x509_csr.sh
PASS: test/rsapss_genpkey_sign_rawin.sh
FAIL: test/rsapss_createak_tls_server.sh
FAIL: test/rsa_pki/rsa_pki.sh
PASS: test/ec_genpkey_check.sh
PASS: test/ec_genpkey_parameters.sh
PASS: test/ecdsa_genpkey_auth.sh
PASS: test/ecdsa_genpkey_sign_rawin.sh
FAIL: test/ecdsa_createak_sign_handle.sh
PASS: test/ecdh_genpkey_keyexch.sh
FAIL: test/ec_createak_x509_csr.sh
PASS: test/ec_genpkey_tls_server.sh
FAIL: test/ec_pki/ec_pki.sh
PASS: test/store_errors.sh
PASS: test/digest.sh
PASS: test/cipher_aes128_ecb.sh
PASS: test/cipher_aes128_cbc.sh
PASS: test/cipher_aes256.sh
PASS: test/cipher_aes256_nopad.sh
FAIL: test/cipher_camellia128.sh
FAIL: test/selftest
============================================================================
Testsuite summary for tpm2-openssl 2.0.0
============================================================================
# TOTAL: 34
# PASS:  19
# SKIP:  0
# XFAIL: 0
# FAIL:  15
# XPASS: 0
# ERROR: 0
============================================================================
See ./test-suite.log
Please report to https://github.com/gotthardp/tpm2-openssl/issues
============================================================================
make[2]: *** [Makefile:1122: test-suite.log] Error 1
make[2]: Leaving directory '/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl'
make[1]: *** [Makefile:1230: check-TESTS] Error 2
make[1]: Leaving directory '/home/levitte/gitwrk/github.com/tpm2-software/_tpm2-openssl'
make: *** [Makefile:1443: check-am] Error 2
```

That's two more tests passing since https://github.com/tpm2-software/tpm2-openssl/issues/5#issuecomment-868252353
Progress! :wink:

**Comment by gotthardp** [comment]
In the past I had bad experiences with `-provider default` without `-propquery ?provider=tpm2`: it creates e.g. two RSA keymgmts and openssl then randomly chooses one, sometimes the wrong one. The `propquery` was a hint to resolve the conflicts between same names. [But the openssl logic might have been improved in the meantime.]

**Comment by levitte** [comment]
> [But the openssl logic might have been improved in the meantime.]

Yeah, there was a few bugs there that made things a bit difficult.  I *think* (can't remember and don't have time to check now) that it does a better job trying all available keymgmts, and also that it has better control over the structures being passed on.  We'll see what my future attempts will lead to.

**Comment by levitte** [comment]
TPM2-openssl turns out to be a good test case for things we're not doing quite right (seems to be small stuff so far) as well as things we might not have communicated very well...

**Comment by levitte** [comment]
I've come to a point where I mainly have failures that I don't know how to deal with:

``` console
$ for x in test/*.log; do tail -1 $x | ( if grep -q '^FAIL'; then tail -20 $x | sed -e "s|^|$x : |"; fi ); done
test/ec_createak_x509_csr.log :   name: 000bb59d9a77dba6befd3e4b60dd41d8b84a6fa481e46790fd74cf5a3273d9bc1cce
test/ec_createak_x509_csr.log :   qualified name: 000b731df07c088db9b34c7852bb7fda1afece6d9213acea83f3549a28339f0badc0
test/ec_createak_x509_csr.log : ++ tpm2_evictcontrol -c ak_rsa.ctx
test/ec_createak_x509_csr.log : ++ cut -d ' ' -f 2
test/ec_createak_x509_csr.log : ++ head -n 1
test/ec_createak_x509_csr.log : WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
test/ec_createak_x509_csr.log : ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
test/ec_createak_x509_csr.log : ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
test/ec_createak_x509_csr.log : ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
test/ec_createak_x509_csr.log : ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
test/ec_createak_x509_csr.log : ERROR: Unable to run tpm2_evictcontrol
test/ec_createak_x509_csr.log : + HANDLE=
test/ec_createak_x509_csr.log : + cat
test/ec_createak_x509_csr.log : + openssl req -provider tpm2 -provider base -new -config testcert.conf -key handle: -out testcsr.pem
test/ec_createak_x509_csr.log : WARNING:esys:src/tss2-esys/api/Esys_ReadPublic.c:320:Esys_ReadPublic_Finish() Received TPM Error 
test/ec_createak_x509_csr.log : ERROR:esys:src/tss2-esys/esys_tr.c:231:Esys_TR_FromTPMPublic_Finish() Error ReadPublic ErrorCode (0x00000184) 
test/ec_createak_x509_csr.log : ERROR:esys:src/tss2-esys/esys_tr.c:321:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x00000184) 
test/ec_createak_x509_csr.log : Could not read private key from handle:
test/ec_createak_x509_csr.log : 80C2E2B62D7F0000:error:4180000C:tpm2::cannot load key::-1:388 tpm:handle(1):value is out of range or is not correct for the context
test/ec_createak_x509_csr.log : FAIL test/ec_createak_x509_csr.sh (exit status: 1)
test/ecdh_create_keyexch_index.log : sym-keybits: 0
test/ecdh_create_keyexch_index.log : x: 5fea730534cbe80fe063d01f5e37b11e6058057476013c9b7d84d6d483fa82b1
test/ecdh_create_keyexch_index.log : y: a8770f1b14c1f022d43e7efca6df170fea588f2183abb8cc51c0b479e393a6e4
test/ecdh_create_keyexch_index.log : ++ tpm2_evictcontrol -c testkey1.ctx
test/ecdh_create_keyexch_index.log : ++ cut -d ' ' -f 2
test/ecdh_create_keyexch_index.log : ++ head -n 1
test/ecdh_create_keyexch_index.log : WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
test/ecdh_create_keyexch_index.log : ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
test/ecdh_create_keyexch_index.log : ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
test/ecdh_create_keyexch_index.log : ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
test/ecdh_create_keyexch_index.log : ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
test/ecdh_create_keyexch_index.log : ERROR: Unable to run tpm2_evictcontrol
test/ecdh_create_keyexch_index.log : + HANDLE=
test/ecdh_create_keyexch_index.log : + openssl pkey -provider tpm2 -in handle: -pubout -out testkey1.pub
test/ecdh_create_keyexch_index.log : WARNING:esys:src/tss2-esys/api/Esys_ReadPublic.c:320:Esys_ReadPublic_Finish() Received TPM Error 
test/ecdh_create_keyexch_index.log : ERROR:esys:src/tss2-esys/esys_tr.c:231:Esys_TR_FromTPMPublic_Finish() Error ReadPublic ErrorCode (0x00000184) 
test/ecdh_create_keyexch_index.log : ERROR:esys:src/tss2-esys/esys_tr.c:321:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x00000184) 
test/ecdh_create_keyexch_index.log : Could not read key from handle:
test/ecdh_create_keyexch_index.log : 8012FCD0727F0000:error:4180000C:tpm2::cannot load key::-1:388 tpm:handle(1):value is out of range or is not correct for the context
test/ecdh_create_keyexch_index.log : FAIL test/ecdh_create_keyexch_index.sh (exit status: 1)
test/ecdsa_createak_sign_handle.log :   name: 000b94b152dc967087f96d153cd0781641a7b95447945d86eaa8958b7450bf499281
test/ecdsa_createak_sign_handle.log :   qualified name: 000b66d364bc634774f4bc53bc8e6a920c30823985329731beb80d2900bf81aaaa36
test/ecdsa_createak_sign_handle.log : ++ tpm2_evictcontrol -c ak_rsa.ctx
test/ecdsa_createak_sign_handle.log : ++ cut -d ' ' -f 2
test/ecdsa_createak_sign_handle.log : ++ head -n 1
test/ecdsa_createak_sign_handle.log : WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
test/ecdsa_createak_sign_handle.log : ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
test/ecdsa_createak_sign_handle.log : ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
test/ecdsa_createak_sign_handle.log : ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
test/ecdsa_createak_sign_handle.log : ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
test/ecdsa_createak_sign_handle.log : ERROR: Unable to run tpm2_evictcontrol
test/ecdsa_createak_sign_handle.log : + HANDLE=
test/ecdsa_createak_sign_handle.log : + openssl pkeyutl -provider tpm2 -provider base -inkey handle: -sign -rawin -in testdata -out testdata.sig
test/ecdsa_createak_sign_handle.log : WARNING:esys:src/tss2-esys/api/Esys_ReadPublic.c:320:Esys_ReadPublic_Finish() Received TPM Error 
test/ecdsa_createak_sign_handle.log : ERROR:esys:src/tss2-esys/esys_tr.c:231:Esys_TR_FromTPMPublic_Finish() Error ReadPublic ErrorCode (0x00000184) 
test/ecdsa_createak_sign_handle.log : ERROR:esys:src/tss2-esys/esys_tr.c:321:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x00000184) 
test/ecdsa_createak_sign_handle.log : Could not read private key from handle:
test/ecdsa_createak_sign_handle.log : 80C228EDD37F0000:error:4180000C:tpm2::cannot load key::-1:388 tpm:handle(1):value is out of range or is not correct for the context
test/ecdsa_createak_sign_handle.log : pkeyutl: Error initializing context
test/ecdsa_createak_sign_handle.log : FAIL test/ecdsa_createak_sign_handle.sh (exit status: 1)
test/rsa_createak_auth.log :   name: 000bf585c7b0eae280cf06e0a6879bc5167cdd20bf6bfc9fb560c321097403397e40
test/rsa_createak_auth.log :   qualified name: 000b6d8eb5c05ef109b4aa7453e153c7e85961791b7343ab4b231372605d3ca02e93
test/rsa_createak_auth.log : ++ tpm2_evictcontrol -c ak_rsa.ctx
test/rsa_createak_auth.log : ++ cut -d ' ' -f 2
test/rsa_createak_auth.log : ++ head -n 1
test/rsa_createak_auth.log : WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
test/rsa_createak_auth.log : ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
test/rsa_createak_auth.log : ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
test/rsa_createak_auth.log : ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
test/rsa_createak_auth.log : ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
test/rsa_createak_auth.log : ERROR: Unable to run tpm2_evictcontrol
test/rsa_createak_auth.log : + HANDLE=
test/rsa_createak_auth.log : + openssl pkeyutl -provider tpm2 -provider base -inkey 'handle:?pass' -sign -rawin -in testdata -passin pass:abc -out testdata.sig
test/rsa_createak_auth.log : WARNING:esys:src/tss2-esys/api/Esys_ReadPublic.c:320:Esys_ReadPublic_Finish() Received TPM Error 
test/rsa_createak_auth.log : ERROR:esys:src/tss2-esys/esys_tr.c:231:Esys_TR_FromTPMPublic_Finish() Error ReadPublic ErrorCode (0x00000184) 
test/rsa_createak_auth.log : ERROR:esys:src/tss2-esys/esys_tr.c:321:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x00000184) 
test/rsa_createak_auth.log : Could not read private key from handle:?pass
test/rsa_createak_auth.log : 8082500D4C7F0000:error:4180000C:tpm2::cannot load key::-1:388 tpm:handle(1):value is out of range or is not correct for the context
test/rsa_createak_auth.log : pkeyutl: Error initializing context
test/rsa_createak_auth.log : FAIL test/rsa_createak_auth.sh (exit status: 1)
test/rsa_createak_sign_handle.log :   name: 000b51d2e4f7e37b02faa68ef16e5ae69a14ae595f535c52dfb09ed63e14bbf05dcf
test/rsa_createak_sign_handle.log :   qualified name: 000b403307bc878069b41fb516c87453762c757f93cd9510bd7e48ba70cf25857029
test/rsa_createak_sign_handle.log : ++ tpm2_evictcontrol -c ak_rsa.ctx
test/rsa_createak_sign_handle.log : ++ cut -d ' ' -f 2
test/rsa_createak_sign_handle.log : ++ head -n 1
test/rsa_createak_sign_handle.log : WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
test/rsa_createak_sign_handle.log : ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
test/rsa_createak_sign_handle.log : ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
test/rsa_createak_sign_handle.log : ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
test/rsa_createak_sign_handle.log : ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
test/rsa_createak_sign_handle.log : ERROR: Unable to run tpm2_evictcontrol
test/rsa_createak_sign_handle.log : + HANDLE=
test/rsa_createak_sign_handle.log : + openssl pkeyutl -provider tpm2 -provider base -inkey handle: -sign -rawin -in testdata -out testdata.sig
test/rsa_createak_sign_handle.log : WARNING:esys:src/tss2-esys/api/Esys_ReadPublic.c:320:Esys_ReadPublic_Finish() Received TPM Error 
test/rsa_createak_sign_handle.log : ERROR:esys:src/tss2-esys/esys_tr.c:231:Esys_TR_FromTPMPublic_Finish() Error ReadPublic ErrorCode (0x00000184) 
test/rsa_createak_sign_handle.log : ERROR:esys:src/tss2-esys/esys_tr.c:321:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x00000184) 
test/rsa_createak_sign_handle.log : Could not read private key from handle:
test/rsa_createak_sign_handle.log : 80822160517F0000:error:4180000C:tpm2::cannot load key::-1:388 tpm:handle(1):value is out of range or is not correct for the context
test/rsa_createak_sign_handle.log : pkeyutl: Error initializing context
test/rsa_createak_sign_handle.log : FAIL test/rsa_createak_sign_handle.sh (exit status: 1)
test/rsa_createak_sign_object.log :   qualified name: 000b5fe683b9a4af9b27a923dc9fa6204b9cfddf6b245c62c72a8ad2a6f4248ccb6b
test/rsa_createak_sign_object.log : ++ tpm2_evictcontrol -c ak_rsa.ctx -o ak_rsa.obj
test/rsa_createak_sign_object.log : ++ cut -d ' ' -f 2
test/rsa_createak_sign_object.log : ++ head -n 1
test/rsa_createak_sign_object.log : WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
test/rsa_createak_sign_object.log : ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
test/rsa_createak_sign_object.log : ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
test/rsa_createak_sign_object.log : ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
test/rsa_createak_sign_object.log : ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
test/rsa_createak_sign_object.log : ERROR: Unable to run tpm2_evictcontrol
test/rsa_createak_sign_object.log : + HANDLE=
test/rsa_createak_sign_object.log : + openssl pkeyutl -provider tpm2 -provider base -inkey object:ak_rsa.obj -sign -rawin -in testdata -out testdata.sig
test/rsa_createak_sign_object.log : Could not open file or uri for loading private key from object:ak_rsa.obj
test/rsa_createak_sign_object.log : 80626A95DD7F0000:error:16000069:STORE routines:ossl_store_get0_loader_int:unregistered scheme:crypto/store/store_register.c:237:scheme=file
test/rsa_createak_sign_object.log : 80626A95DD7F0000:error:80000002:system library:file_open:No such file or directory:providers/implementations/storemgmt/file_store.c:269:calling stat(object:ak_rsa.obj)
test/rsa_createak_sign_object.log : 80626A95DD7F0000:error:16000069:STORE routines:ossl_store_get0_loader_int:unregistered scheme:crypto/store/store_register.c:237:scheme=object
test/rsa_createak_sign_object.log : 80626A95DD7F0000:error:80000002:system library:BIO_new_file:No such file or directory:crypto/bio/bss_file.c:67:calling fopen(ak_rsa.obj, rb)
test/rsa_createak_sign_object.log : 80626A95DD7F0000:error:10000080:BIO routines:BIO_new_file:no such file:crypto/bio/bss_file.c:75:
test/rsa_createak_sign_object.log : pkeyutl: Error initializing context
test/rsa_createak_sign_object.log : FAIL test/rsa_createak_sign_object.sh (exit status: 1)
test/rsa_createak_x509_csr.log :   name: 000b8160a1644f6108faa3832c0738514d4fe7aae4daa5a5fe379bc05f672eada8f6
test/rsa_createak_x509_csr.log :   qualified name: 000b4769393565262f45303cb7c2d5f37cd8b317c1c33c28b9c0b3fe47351c2f5799
test/rsa_createak_x509_csr.log : ++ tpm2_evictcontrol -c ak_rsa.ctx
test/rsa_createak_x509_csr.log : ++ cut -d ' ' -f 2
test/rsa_createak_x509_csr.log : ++ head -n 1
test/rsa_createak_x509_csr.log : WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
test/rsa_createak_x509_csr.log : ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
test/rsa_createak_x509_csr.log : ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
test/rsa_createak_x509_csr.log : ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
test/rsa_createak_x509_csr.log : ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
test/rsa_createak_x509_csr.log : ERROR: Unable to run tpm2_evictcontrol
test/rsa_createak_x509_csr.log : + HANDLE=
test/rsa_createak_x509_csr.log : + cat
test/rsa_createak_x509_csr.log : + openssl req -provider tpm2 -new -config testcert.conf -key handle: -out testcsr.pem
test/rsa_createak_x509_csr.log : WARNING:esys:src/tss2-esys/api/Esys_ReadPublic.c:320:Esys_ReadPublic_Finish() Received TPM Error 
test/rsa_createak_x509_csr.log : ERROR:esys:src/tss2-esys/esys_tr.c:231:Esys_TR_FromTPMPublic_Finish() Error ReadPublic ErrorCode (0x00000184) 
test/rsa_createak_x509_csr.log : ERROR:esys:src/tss2-esys/esys_tr.c:321:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x00000184) 
test/rsa_createak_x509_csr.log : Could not read private key from handle:
test/rsa_createak_x509_csr.log : 8022803C6C7F0000:error:4180000C:tpm2::cannot load key::-1:388 tpm:handle(1):value is out of range or is not correct for the context
test/rsa_createak_x509_csr.log : FAIL test/rsa_createak_x509_csr.sh (exit status: 1)
test/rsa_create_decrypt.log : rsa: b1b00dc19cb05742b4c3bd81657f94cd5e2cea2d0cac701625c1fc253f865ba38bc864251edf26fb82e8e2b75905448a3568a604a99c37d284e5fe4ffc624b625da5591362bc411b24eaba24e06ee99506ae18143e0db4b00e850ba877104e16d4377f448ec25cc1bbabb1244a4d38a43af74950af8957be778186d4f87fce6fb8ff44abdfecc99fb8ea21fb42455ef9d4524b54ad8519afa825c9c8fd312c1b0f516d3accc2d7898ab7bb4c8a6c4c95cc97a14f22a0119f365f76f7d0ab22c6528008e5d7d71706c925fe68553fcf8bfb489177add844f4d26cc466b38f5d917af6fa3d5c6e05e4367980186a8f607dd1be2adf4b69749bdaf4ce91c9a1b5d9
test/rsa_create_decrypt.log : + tpm2_load -C primary.ctx -u key.pub -r key.priv -c testkey.ctx
test/rsa_create_decrypt.log : name: 000b4e9d64c0369417e0b6c1956219c09054e7e2cc231908bcd1abce06dd6196094e
test/rsa_create_decrypt.log : ++ tpm2_evictcontrol -c testkey.ctx
test/rsa_create_decrypt.log : ++ cut -d ' ' -f 2
test/rsa_create_decrypt.log : ++ head -n 1
test/rsa_create_decrypt.log : WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
test/rsa_create_decrypt.log : ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
test/rsa_create_decrypt.log : ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
test/rsa_create_decrypt.log : ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
test/rsa_create_decrypt.log : ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
test/rsa_create_decrypt.log : ERROR: Unable to run tpm2_evictcontrol
test/rsa_create_decrypt.log : + HANDLE=
test/rsa_create_decrypt.log : + openssl pkey -provider tpm2 -in handle: -pubout -out testkey.pub
test/rsa_create_decrypt.log : WARNING:esys:src/tss2-esys/api/Esys_ReadPublic.c:320:Esys_ReadPublic_Finish() Received TPM Error 
test/rsa_create_decrypt.log : ERROR:esys:src/tss2-esys/esys_tr.c:231:Esys_TR_FromTPMPublic_Finish() Error ReadPublic ErrorCode (0x00000184) 
test/rsa_create_decrypt.log : ERROR:esys:src/tss2-esys/esys_tr.c:321:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x00000184) 
test/rsa_create_decrypt.log : Could not read key from handle:
test/rsa_create_decrypt.log : 80E21F18977F0000:error:4180000C:tpm2::cannot load key::-1:388 tpm:handle(1):value is out of range or is not correct for the context
test/rsa_create_decrypt.log : FAIL test/rsa_create_decrypt.sh (exit status: 1)
test/rsa_genpkey_auth_parent.log :   value: aes
test/rsa_genpkey_auth_parent.log :   raw: 0x6
test/rsa_genpkey_auth_parent.log : sym-mode:
test/rsa_genpkey_auth_parent.log :   value: cfb
test/rsa_genpkey_auth_parent.log :   raw: 0x43
test/rsa_genpkey_auth_parent.log : sym-keybits: 128
test/rsa_genpkey_auth_parent.log : rsa: d796316db9b1116c18f120b444b378705d7859030887c7a5572472b33c0d83d780e86b2b15fbd6c7ff1f9658973433a1498313c83efe39ec28d8972dcbe32941f628774a583b694650f7ca3b734736e26346ae3eec311d136d3bd6c35ecf96609ca1c6a0246c987594791569d02a3ea6927c993bcf7dd7eba0f55612d299bae2e43b78555bbba2c6a48e2ef380f92fe01aff42e750c45b0e036cb200576d0e47699f2bece26b60fc0f3645ab1c7f0135994b2bf8427e3665e6f999df3e04ed8b48ac55c1d0dccd5f8dc51ca26980a18c0eb9da608d7538ea35eec25852f28551e34b26d6448ce0c61fdc675a03babc7cfbb5c880bb8774e6f4aa20bd911a127f
test/rsa_genpkey_auth_parent.log : ++ tpm2_evictcontrol -c parent.ctx
test/rsa_genpkey_auth_parent.log : ++ cut -d ' ' -f 2
test/rsa_genpkey_auth_parent.log : ++ head -n 1
test/rsa_genpkey_auth_parent.log : WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
test/rsa_genpkey_auth_parent.log : ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
test/rsa_genpkey_auth_parent.log : ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
test/rsa_genpkey_auth_parent.log : ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
test/rsa_genpkey_auth_parent.log : ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
test/rsa_genpkey_auth_parent.log : ERROR: Unable to run tpm2_evictcontrol
test/rsa_genpkey_auth_parent.log : + HANDLE=
test/rsa_genpkey_auth_parent.log : + openssl genpkey -provider tpm2 -provider base -algorithm RSA -out testkey.priv -pkeyopt parent: -pkeyopt user-auth:abc -pkeyopt bits:1024
test/rsa_genpkey_auth_parent.log : genpkey: Error setting parent: parameter:
test/rsa_genpkey_auth_parent.log : FAIL test/rsa_genpkey_auth_parent.sh (exit status: 1)
test/rsa_genpkey_x509_cmp.log :         08:f0:9c:27:cf:01:50:ee:df:b3:47:4e:59:d4:41:62:ee:fb:
test/rsa_genpkey_x509_cmp.log :         2c:e7:c7:76:55:d2:3a:a9:87:cb:b4:e6:34:71:74:aa:2c:cd:
test/rsa_genpkey_x509_cmp.log :         30:84:52:bc:ea:85:a1:52:88:7c:40:d7:6b:75:71:6c:d6:bd:
test/rsa_genpkey_x509_cmp.log :         f0:69:9e:f1:dd:70:70:82:b8:72:af:ad:ed:3a:1b:67:b0:6f:
test/rsa_genpkey_x509_cmp.log :         1c:ef:48:9d:60:d3:22:75:3d:dd:27:d6:a4:56:91:bf:db:9e:
test/rsa_genpkey_x509_cmp.log :         1e:f7:61:0a:3a:f4:54:69:77:0c:46:30:e0:12:0d:19:9b:99:
test/rsa_genpkey_x509_cmp.log :         94:ea:21:c1:a3:a6:f3:69:b4:cc:d2:74:ce:b7:a4:a5:98:46:
test/rsa_genpkey_x509_cmp.log :         fe:82:eb:34:51:e0:ca:a9:48:9d:14:b6:73:76:1e:c8:94:80:
test/rsa_genpkey_x509_cmp.log :         43:57:93:02:f8:58:d2:14:31:b5:68:4c:e8:b1:ab:b0:a7:8e:
test/rsa_genpkey_x509_cmp.log :         fc:77:91:e3:0b:d0:49:04:97:51:88:3f:9c:3f:e8:93:12:95:
test/rsa_genpkey_x509_cmp.log :         84:b1:fb:f5:5b:ee:ca:a9:24:82:32:7f:37:88:43:d0:da:02:
test/rsa_genpkey_x509_cmp.log :         5b:71:e1:d6:fc:1a:61:45:a5:34:62:fb:91:07:57:00:7e:4d:
test/rsa_genpkey_x509_cmp.log :         6d:fc:f8:3d:dc:91:8a:f2:29:0f:0e:7b:85:bf:11:60:b8:c6:
test/rsa_genpkey_x509_cmp.log :         7c:3c:30:6b:3a:ac:1f:29:33:73:5f:a7:f9:c2:d9:0f:18:56:
test/rsa_genpkey_x509_cmp.log :         f7:ec:78:fe
test/rsa_genpkey_x509_cmp.log : + openssl genrsa -out test-server-key.pem 2048
test/rsa_genpkey_x509_cmp.log : + openssl req -provider tpm2 -provider default -x509 -sha256 -nodes -subj /C=GB/CN=server.example.com -extensions usr_cert -CAkey test-ca-key.pem -CA test-ca-cert.pem -key test-server-key.pem -out test-server-cert.pem
test/rsa_genpkey_x509_cmp.log : -----
test/rsa_genpkey_x509_cmp.log : Could not read issuer private key from test-ca-key.pem
test/rsa_genpkey_x509_cmp.log : FAIL test/rsa_genpkey_x509_cmp.sh (exit status: 1)
test/rsa_genrsa_check.log : + cat pubkey.pem
test/rsa_genrsa_check.log : + openssl pkey -provider tpm2 -provider base -pubout -outform der -out testkey.der
test/rsa_genrsa_check.log : + openssl rsa -pubin -inform der -in testkey.der -modulus -noout
test/rsa_genrsa_check.log : Modulus=B0B4E80867CFB2D2847776F1D43B0940738456814C169E1370D7D76B543D3124C1ACB713AF2EB26F5DF683AC08FC038FCB7AEAB9B0D89FEFB45D62C4F6F28DCE6F885D7A3C5A4AFE87379DF634338E56DADF1A53F32FBC9C48E3723DADA4019F6B386AF51138E8BC9E7EB9B60992E93BBEE4728245CC1583626109F1B67274EB
test/rsa_genrsa_check.log : + openssl rsa -pubin -inform der -in testkey.der -text -noout
test/rsa_genrsa_check.log : Public-Key: (1024 bit)
test/rsa_genrsa_check.log : Modulus:
test/rsa_genrsa_check.log :     00:b0:b4:e8:08:67:cf:b2:d2:84:77:76:f1:d4:3b:
test/rsa_genrsa_check.log :     09:40:73:84:56:81:4c:16:9e:13:70:d7:d7:6b:54:
test/rsa_genrsa_check.log :     3d:31:24:c1:ac:b7:13:af:2e:b2:6f:5d:f6:83:ac:
test/rsa_genrsa_check.log :     08:fc:03:8f:cb:7a:ea:b9:b0:d8:9f:ef:b4:5d:62:
test/rsa_genrsa_check.log :     c4:f6:f2:8d:ce:6f:88:5d:7a:3c:5a:4a:fe:87:37:
test/rsa_genrsa_check.log :     9d:f6:34:33:8e:56:da:df:1a:53:f3:2f:bc:9c:48:
test/rsa_genrsa_check.log :     e3:72:3d:ad:a4:01:9f:6b:38:6a:f5:11:38:e8:bc:
test/rsa_genrsa_check.log :     9e:7e:b9:b6:09:92:e9:3b:be:e4:72:82:45:cc:15:
test/rsa_genrsa_check.log :     83:62:61:09:f1:b6:72:74:eb
test/rsa_genrsa_check.log : Exponent: 65537 (0x10001)
test/rsa_genrsa_check.log : + openssl rsa -provider tpm2 -provider default -propquery '?provider=tpm2,tpm2.store=no' -in pubkey.der -inform der -RSAPublicKey_out -out testrsa.pem
test/rsa_genrsa_check.log : Could not read private key from pubkey.der
test/rsa_genrsa_check.log : FAIL test/rsa_genrsa_check.sh (exit status: 1)
test/rsapss_createak_tls_server.log :   name: 000bb85675c92f68caabe6007c7b41a595a3dc22a9e182590025f3517bf1b3b18877
test/rsapss_createak_tls_server.log :   qualified name: 000b4bb9a328d14a52fed3aa58a2704c4ee02a15c6637ca0fc869b8a015901153107
test/rsapss_createak_tls_server.log : ++ tpm2_evictcontrol -c ak_rsa.ctx
test/rsapss_createak_tls_server.log : ++ cut -d ' ' -f 2
test/rsapss_createak_tls_server.log : ++ head -n 1
test/rsapss_createak_tls_server.log : WARNING:esys:src/tss2-esys/api/Esys_EvictControl.c:330:Esys_EvictControl_Finish() Received TPM Error 
test/rsapss_createak_tls_server.log : ERROR:esys:src/tss2-esys/api/Esys_EvictControl.c:114:Esys_EvictControl() Esys Finish ErrorCode (0x0000014b) 
test/rsapss_createak_tls_server.log : ERROR: Esys_EvictControl(0x14B) - tpm:error(2.0): insufficient space for NV allocation
test/rsapss_createak_tls_server.log : ERROR:esys:src/tss2-esys/esys_tr.c:357:Esys_TR_Close() Error: Esys handle does not exist (70018). 
test/rsapss_createak_tls_server.log : ERROR: Esys_TR_Close(0x70018) - esapi:The ESYS_TR resource object is bad
test/rsapss_createak_tls_server.log : ERROR: Unable to run tpm2_evictcontrol
test/rsapss_createak_tls_server.log : + HANDLE=
test/rsapss_createak_tls_server.log : + cat
test/rsapss_createak_tls_server.log : + openssl req -provider tpm2 -provider base -x509 -config testcert.conf -key handle: -out testcert.pem
test/rsapss_createak_tls_server.log : WARNING:esys:src/tss2-esys/api/Esys_ReadPublic.c:320:Esys_ReadPublic_Finish() Received TPM Error 
test/rsapss_createak_tls_server.log : ERROR:esys:src/tss2-esys/esys_tr.c:231:Esys_TR_FromTPMPublic_Finish() Error ReadPublic ErrorCode (0x00000184) 
test/rsapss_createak_tls_server.log : ERROR:esys:src/tss2-esys/esys_tr.c:321:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x00000184) 
test/rsapss_createak_tls_server.log : Could not read private key from handle:
test/rsapss_createak_tls_server.log : 8062D07AED7F0000:error:4180000C:tpm2::cannot load key::-1:388 tpm:handle(1):value is out of range or is not correct for the context
test/rsapss_createak_tls_server.log : FAIL test/rsapss_createak_tls_server.sh (exit status: 1)
```

@gotthardp, if you know what I can do about the `insufficient space for NV allocation` issue, I'd be grateful.

**Comment by levitte** [comment]
I'll check what I can do about the `rsa_genrsa_check` failure on Monday...

**Comment by gotthardp** [comment]
@levitte The "insufficient space for NV allocation" is caused by previous errors-- the tests were stopped before cleanup, so the non-volatile (NV) memory of the TPM got full.

In the directory where you start the TPM simulator should be a file named `NVChip`. Delete it and restart both simulator and resource manager.

**Comment by levitte** [comment]
> In the directory where you start the TPM simulator should be a file named `NVChip`. Delete it and restart both simulator and resource manager.

... except I'm not running a simulator :wink:

**Comment by levitte** [comment]
> @levitte The "insufficient space for NV allocation" is caused by previous errors-- the tests were stopped before cleanup, so the non-volatile (NV) memory of the TPM got full.

Thanks...  I'll look more thoroughly at the logs

**Comment by gotthardp** [comment]
If you have a true TPM, then you have to remove the trash manually:
```
$ tpm2_getcap handles-persistent
0x8.......
$ tpm2_evictcontrol -c 0x8.......
```

**Comment by levitte** [comment]
Thank you!  Avicting those made quite a difference!

```
============================================================================
Testsuite summary for tpm2-openssl 2.0.0
============================================================================
# TOTAL: 35
# PASS:  27
# SKIP:  1
# XFAIL: 0
# FAIL:  7
# XPASS: 0
# ERROR: 0
```

**Comment by levitte** [comment]
...  and now I see how I've ended up with a long list of persistent handles.  That may happen for any test that fails...

**Comment by gotthardp** [comment]
The situation is much less desperate than before. I fixed some more bugs (not STORE related) and added few more `-provider base`, so now only one test fails-- the CMP.
```
FAIL: test/rsa_genpkey_x509_cmp.sh
```
The CMP has problems reading a TPM and a non-TPM key in one command. That behaviour will likely be impacted by removal of the file STORE re-implementation.

I tried to remove the file STORE as suggested in #8, but it makes things much worse: 16 failures. Loading a certificate fails even with the base provider:
```
$ openssl pkey -provider tpm2 -provider base -in testkey1.priv -pubout -out testkey1.pub
PROVIDER INIT
C0D7D579C37F0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported::0:
DER DECODER DECODE
TSS2 DECODER DECODE 0x0
Could not read key from testkey1.priv
PROVIDER TEARDOWN
```
There is something very wrong-- the decoders are unable to read the BIO data. I really wonder how you managed to pass so many tests even without the file store. I will do some more debugging tomorrow.

**Comment by levitte** [comment]
> The CMP has problems reading a TPM and a non-TPM key in one command.

Yes, this has to be recognised.  I've played a bit with OpenSSL to tighten the coupling between decoders/encoders and keymgmts, that will certainly help.

> Loading a certificate fails even with the base provider:

I've seen this too.

There are several things interacting at once here:

1.  Your keymgmt import / export functions are a little untight...  only checking for `OSSL_KEYMGMT_SELECT_PUBLIC_KEY` isn't enough, because that may come *combined* with `OSSL_KEYMGMT_SELECT_PRIVATE_KEY`, so on import, you end up creating a public key even though a private key was passed to you, and on export, you end up exporting a public key even though a private key was asked for.
    You shouldn't even need an import function in your keymgmts.  As far as I can tell, TPM keys are kinda sorta an HSM, at least toward OpenSSL.  Your source for TPM based keys can then only be TPM.  That makes any import invalid.
2.  The not so tight coupling between decoders and keymgmts that I'm kinda mentioning above.  OpenSSL doesn't really try to  use the keymgmt associated with the decoder that just decoded something, we just fetch one and hope for luck.
    So what I've seen happen is that your DER->key dcoder did decode the key properly, but then OpenSSL's RSA keymgmt was chosen, and we get the situation where your decoder exports *the public part* of the key, which OpenSSL's RSA keymgmt imports, but then the STORE library refuses it, because the loader returned a public key when a private key was expected.
3.  Also, the selection bits aren't passed around properly by OpenSSL.  `-pubout` would mean that `OSSL_KEYMGMT_SELECT_PRIVATE_KEY` shouldn't be set in the selection that is passed to the keymgmt or the decoder's export function...  and yet, we pass a hard coded `OSSL_KEYMGMT_SELECT_ALL`, so there's no way for the provider code to know what was actually desired.

> I really wonder how you managed to pass so many tests even without the file store.

By fixing the OpenSSL side of things as well.

I haven't submitted everything I've done yet.  Will put them together in well enough formed PRs tomorrow.  Today is Sunday, I'm off to do other things :wink:

Mind you, I played a bit yesterday and triggered other bugs, so for the moment, this is where I'm at:

```
PASS: test/list.sh
PASS: test/rand.sh
FAIL: test/rsa_genrsa_check.sh
PASS: test/rsa_genpkey_sign.sh
PASS: test/rsa_genpkey_sign_rawin.sh
PASS: test/rsa_genpkey_auth_parent.sh
PASS: test/rsa_createak_auth.sh
PASS: test/rsa_createak_sign_object.sh
PASS: test/rsa_createak_sign_handle.sh
PASS: test/rsa_create_decrypt.sh
PASS: test/rsa_genpkey_x509_cert.sh
FAIL: test/rsa_genpkey_x509_cmp.sh
FAIL: test/rsa_genpkey_tls_server.sh
PASS: test/rsa_createak_x509_csr.sh
PASS: test/rsapss_genpkey_sign_rawin.sh
FAIL: test/rsapss_createak_tls_server.sh
FAIL: test/rsa_pki/rsa_pki.sh
PASS: test/ec_genpkey_check.sh
PASS: test/ec_genpkey_parameters.sh
PASS: test/ecdsa_genpkey_auth.sh
PASS: test/ecdsa_genpkey_sign_rawin.sh
PASS: test/ecdsa_createak_sign_handle.sh
PASS: test/ecdh_genpkey_keyexch.sh
PASS: test/ecdh_create_keyexch_index.sh
PASS: test/ec_createak_x509_csr.sh
FAIL: test/ec_genpkey_tls_server.sh
FAIL: test/ec_pki/ec_pki.sh
PASS: test/store_errors.sh
PASS: test/digest.sh
PASS: test/cipher_aes128_ecb.sh
PASS: test/cipher_aes128_cbc.sh
PASS: test/cipher_aes256.sh
PASS: test/cipher_aes256_nopad.sh
SKIP: test/cipher_camellia128.sh
FAIL: test/selftest
============================================================================
Testsuite summary for tpm2-openssl 2.0.0
============================================================================
# TOTAL: 35
# PASS:  26
# SKIP:  1
# XFAIL: 0
# FAIL:  8
# XPASS: 0
# ERROR: 0
============================================================================
See ./test-suite.log
Please report to https://github.com/gotthardp/tpm2-openssl/issues
============================================================================
```

**Comment by levitte** [comment]
BTW, I aim to remove all needs for command line propqueries.  They shouldn't be necessary at all as soon as the decoder-keymgmt coupling is tightened.  Stay tuned.

**Comment by levitte** [comment]
Oh, right, another thing I found is that OpenSSL's PEM->DER decoder is a bit too general...  it treats TSS2 as a key type with a type specific structure, i.e. makes assumptions about the PEM name that doesn't quite correspond to what tpm2-openssl does.  So I've played with modifying OpenSSL's PEM->DER decoder to simply refuse to pass on the DER decoding of PEM names it doesn't know.

**Comment by gotthardp** [comment]
Thanks for your help.

> You shouldn't even need an import function in your keymgmts. As far as I can tell, TPM keys are kinda sorta an HSM, at least toward OpenSSL. Your source for TPM based keys can then only be TPM. That makes any import invalid.

The import was implemented to facilitate key checks, which are done by ECDH or X.509 related OpenSSL functions. The functions sometimes load both a public and a private key to check that they match. ...but you can convince OpenSSL to use the default key management for it as well.

Another use-case for the import functions is a true import of a private key to the TPM. Although this is not fully implemented right now, one could take an externally generated private key (or even public key) and import it into the TPM for later use.

**Comment by levitte** [comment]
> Another use-case for the import functions is a true import of a private key to the TPM.

Yeah I can see that.  There isn't any explicit support for that in OpenSSL for the moment, though

**Comment by levitte** [comment]
The OpenSSL changes that I was talking about are now PRs:

- openssl/openssl#15930
- openssl/openssl#15933
- openssl/openssl#15934

I have a couple of others in line as well, but these three are very much a reason that so many of the tpm2-openssl tests are working for me.

**Comment by levitte** [comment]
With the 3 OpenSSL PRs above applied on my OpenSSL 3 installation, testing #11 gets me this:

```
PASS: test/list.sh
PASS: test/rand.sh
FAIL: test/rsa_genrsa_check.sh
PASS: test/rsa_genpkey_sign.sh
PASS: test/rsa_genpkey_sign_rawin.sh
PASS: test/rsa_genpkey_auth_parent.sh
PASS: test/rsa_createak_auth.sh
PASS: test/rsa_createak_sign_object.sh
PASS: test/rsa_createak_sign_handle.sh
PASS: test/rsa_create_decrypt.sh
PASS: test/rsa_genpkey_x509_cert.sh
FAIL: test/rsa_genpkey_x509_cmp.sh
FAIL: test/rsa_genpkey_tls_server.sh
PASS: test/rsa_createak_x509_csr.sh
PASS: test/rsapss_genpkey_sign_rawin.sh
FAIL: test/rsapss_createak_tls_server.sh
PASS: test/rsa_pki/rsa_pki.sh
PASS: test/ec_genpkey_check.sh
PASS: test/ec_genpkey_parameters.sh
PASS: test/ecdsa_genpkey_auth.sh
PASS: test/ecdsa_genpkey_sign_rawin.sh
PASS: test/ecdsa_createak_sign_handle.sh
PASS: test/ecdh_genpkey_keyexch.sh
PASS: test/ecdh_create_keyexch_index.sh
PASS: test/ec_createak_x509_csr.sh
FAIL: test/ec_genpkey_tls_server.sh
FAIL: test/ec_pki/ec_pki.sh
PASS: test/store_errors.sh
PASS: test/digest.sh
PASS: test/cipher_aes128_ecb.sh
PASS: test/cipher_aes128_cbc.sh
PASS: test/cipher_aes256.sh
PASS: test/cipher_aes256_nopad.sh
SKIP: test/cipher_camellia128.sh
FAIL: test/selftest
============================================================================
Testsuite summary for tpm2-openssl 2.0.0
============================================================================
# TOTAL: 35
# PASS:  27
# SKIP:  1
# XFAIL: 0
# FAIL:  7
# XPASS: 0
# ERROR: 0
============================================================================
See ./test-suite.log
Please report to https://github.com/gotthardp/tpm2-openssl/issues
============================================================================
```

**Comment by levitte** [comment]
#13 fixes the failing `test/ec_pki/ec_pki.sh`, so I'm now down to this:

```
============================================================================
Testsuite summary for tpm2-openssl 2.0.0
============================================================================
# TOTAL: 35
# PASS:  28
# SKIP:  1
# XFAIL: 0
# FAIL:  6
# XPASS: 0
# ERROR: 0
============================================================================
See ./test-suite.log
Please report to https://github.com/gotthardp/tpm2-openssl/issues
============================================================================
```

The next big problem can be summarized like this:

```
test/ec_genpkey_tls_server.log : 80C2913AF97F0000:error:05800075:x509 certificate routines:X509_check_private_key:unknown key type:crypto/x509/x509_cmp.c:410:
test/rsa_genpkey_tls_server.log : 80427CF19F7F0000:error:05800075:x509 certificate routines:X509_check_private_key:unknown key type:crypto/x509/x509_cmp.c:410:
test/rsa_genpkey_x509_cmp.log : 80228835BC7F0000:error:05800075:x509 certificate routines:X509_check_private_key:unknown key type:crypto/x509/x509_cmp.c:410:
test/rsa_genpkey_x509_cmp.log : 80228835BC7F0000:error:05800075:x509 certificate routines:X509_check_private_key:unknown key type:crypto/x509/x509_cmp.c:410:
```

I believe that's the problem you described, uhm, somewhere, and that @mouse07410 describes in openssl/openssl#15927
That'll be my next focus.

-----

Apart from that, I have this:

```
test/rsa_genrsa_check.log : + openssl rsa -provider tpm2 -provider base -in pubkey.der -inform der -RSAPublicKey_out -out testrsa.pem
test/rsa_genrsa_check.log : Could not read private key from pubkey.der
```

That's a different issue, which I've described in openssl/openssl#15907

Both of these remaining problems are in OpenSSL proper, not in tpm2-openssl

**Comment by gotthardp** [comment]
Thank you! All PR's are now merged.

The automated tests are now at:
```
# TOTAL: 35
# PASS:  30
# SKIP:  0
# XFAIL: 0
# FAIL:  5
# XPASS: 0
# ERROR: 0
```
Your `test/selftest` failure is surprising; it works for me and there is not much what could be wrong. Could you send me your `test/selftest.log` please? No hurry, I am just curious.

**Comment by levitte** [comment]
> Your `test/selftest` failure is surprising

Er, it now passes, so never mind.

**Comment by levitte** [comment]
There's a new PR in OpenSSL that I hope will clear the confusion about where key data is held.  Basically, this off-loads the final object (such as keys) decoding back to libcrypto.

- openssl/openssl#15981

It's possible, though, that there's still going to be a need to be able to export private keys.  I'd suggest having a TPM2 specific export parameter that is a reference of some sort to the secret parts (a handle, for example).

**Comment by levitte** [comment]
I have a strange kind of failure:

```
+ openssl req -provider tpm2 -provider default -x509 -newkey rsa:2048 -sha256 -nodes -subj /C=GB/CN=client.example.com -extensions usr_cert -CAkey test-ca-key.pem -CA test-ca-cert.pem -keyout test-client-key.pem -out test-client-cert.pem
PROVIDER INIT
DIGEST SHA1 GET_PARAMS [ blocksize size xof algid-absent ]
DIGEST SHA256 GET_PARAMS [ blocksize size xof algid-absent ]
RSA GEN INIT rsa 3
RSA GEN_SET_PARAMS [ bits ]
RAND NEW
RAND GET_CTX_PARAMS [ state ]
Warning: generating random key material may take a long time
if the system has a poor entropy source
RSA GEN 2048 bits
RSA GEN parent: primary 0x40000001
RSA GET_PARAMS [ bits security-bits max-size ]
RSA CLEANUP
ENCODER tss PrivateKeyInfo/der DOES_SELECTION 0x87
ENCODER tss PrivateKeyInfo/pem DOES_SELECTION 0x87
ENCODER rsa pkcs1/der DOES_SELECTION 0x87
ENCODER rsa pkcs1/der DOES_SELECTION 0x87
ENCODER rsa pkcs1/der DOES_SELECTION 0x87
ENCODER rsa pkcs1/der DOES_SELECTION 0x87
ENCODER rsa pkcs1/pem DOES_SELECTION 0x87
ENCODER rsa pkcs1/pem DOES_SELECTION 0x87
ENCODER rsa pkcs1/pem DOES_SELECTION 0x87
ENCODER rsa pkcs1/pem DOES_SELECTION 0x87
ENCODER rsa SubjectPublicKeyInfo/der DOES_SELECTION 0x87
ENCODER rsa SubjectPublicKeyInfo/der DOES_SELECTION 0x87
ENCODER rsa SubjectPublicKeyInfo/der DOES_SELECTION 0x87
ENCODER rsa SubjectPublicKeyInfo/der DOES_SELECTION 0x87
ENCODER rsa SubjectPublicKeyInfo/pem DOES_SELECTION 0x87
ENCODER rsa SubjectPublicKeyInfo/pem DOES_SELECTION 0x87
ENCODER rsa SubjectPublicKeyInfo/pem DOES_SELECTION 0x87
ENCODER rsa SubjectPublicKeyInfo/pem DOES_SELECTION 0x87
ENCODER tss PrivateKeyInfo/pem ENCODE 0x87
-----
DER DECODER DECODE
DER DECODER DECODE
TSS2 DECODER DECODE 0x87
TSS2 DECODER LOAD parent: primary 0x40000001
TSS2 DECODER DECODE 0x87
TSS2 DECODER LOAD parent: primary 0x40000001
WARNING:esys:src/tss2-esys/api/Esys_Load.c:324:Esys_Load_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_Load.c:112:Esys_Load() Esys Finish ErrorCode (0x00000902) 
TSS2 DECODER DECODE found (null)
80824A8EBD7F0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151:
DER DECODER DECODE
Could not read issuer private key from test-ca-key.pem
RSA FREE
RAND FREE
PROVIDER TEARDOWN
FAIL test/rsa_genpkey_x509_cmp.sh (exit status: 1)
```

The strange part is this:

```
WARNING:esys:src/tss2-esys/api/Esys_Load.c:324:Esys_Load_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_Load.c:112:Esys_Load() Esys Finish ErrorCode (0x00000902) 
```

and it happens here:

https://github.com/tpm2-software/tpm2-openssl/blob/c98d408d4a8dfeb60dac758425ebe2478817a9da/src/tpm2-provider-decoder-tss2.c#L83-L85

The error code, decoded with `Tss2_RC_Decode()`, is:

```
(gdb) p (const char *)Tss2_RC_Decode(0x902)
$1 = 0x55555571be00 "tpm:warn(2.0): out of memory for object contexts"
```

Unfortunately, I have no idea what to do about that.

This is current master, and openssl master with openssl/openssl#15981 merged in

**Comment by gotthardp** [comment]
The TPM has limited memory can can only store a certain number of objects at the same time-- not only keys, but also sequence for digest calculation. I saw "out of memory" messages in situations when somebody did load "too many keys" and/or started "too many" hash operations, etc. How much is "too many" depends on the TPM chip. CMP is one example of such resource intensive application (see #4) because for CMP Key-Update-Request it wants to loads two private keys, two certificates (and corresponding public keys), plus it starts digest to verify the cert... and that's too many for a TPM to handle. The CMP test thus may fail on a real TPM.

HOWEVER-- in your case, it's just a CSR preparation, not the CMP operation itself. That is not so resource intensive, so it should work. It could be because some FREE operation (that also clears TPM resources) is called too late and there exist too many objects in parallel. Bit it could also be something completely different-- I will try to debug this, but not until Wednesday next week unfortunately as I am about to leave for a short vacation.

**Comment by levitte** [comment]
This is my latest test-suite.log.  The failures are all because of out of memory.  I hope that helps figuring things out.

[test-suite.log](https://github.com/tpm2-software/tpm2-openssl/files/6757884/test-suite.log)

**Comment by gotthardp** [comment]
Hmmm. For now the automated test suite still reports something wrong with export import. It appears to be trying to export a provate key. Are there really no other changes required on the openssl side?

For example, rsa_genpkey_tls_server.log:
```
+ openssl s_server -provider tpm2 -provider default -accept 4443 -www -key testkey.pem -cert testcert.pem
+ sleep 2
PROVIDER INIT
DER DECODER DECODE
DER DECODER DECODE
TSS2 DECODER DECODE 0x87
TSS2 DECODER LOAD parent: primary 0x40000001
TSS2 DECODER DECODE 0x87
TSS2 DECODER LOAD parent: primary 0x40000001
TSS2 DECODER DECODE found RSA
RSA LOAD
RSA GET_PARAMS [ bits security-bits max-size ]
RSA HAS 1
RAND NEW
RAND NEW
RAND GET_CTX_PARAMS [ max_request ]
RAND GENERATE
RAND NEW
RAND GET_CTX_PARAMS [ max_request ]
RAND GENERATE
RAND GET_CTX_PARAMS [ max_request ]
RAND GENERATE
RAND GET_CTX_PARAMS [ max_request ]
RAND GENERATE
DER DECODER DECODE
TSS2 DECODER DECODE 0x84
TSS2 DECODER DECODE 0x84
DER DECODER DECODE
Using default temp DH parameters
DIGEST NEW
DIGEST INIT
DIGEST UPDATE
DIGEST FINAL
DIGEST FREE
RSA NEW
RSA IMPORT [ n e ]
RSA FREE
RSA EXPORT 87
error setting private key
C03711FB367F0000:error:05800075:x509 certificate routines:X509_check_private_key:unknown key type::0:
RSA FREE
RAND FREE
RAND FREE
RAND FREE
PROVIDER TEARDOWN
```

**Comment by levitte** [comment]
Yes, export import is a problem with how tpm2-openssl currently works.  Unfortunately, I seem to be hitting Tss2 error 0x902 before getting there 😖 

So it sounds like I should try with the simulator instead, and see where I get with it...

**Comment by gotthardp** [comment]
I was able to reproduce one tss memory error and made a fix https://github.com/tpm2-software/tpm2-openssl/commit/cb0a070b0dd93f13429e82ee60b41c1cc27c6f90. It may improve the situation also for you.

**Comment by levitte** [comment]
I'll try it out

**Comment by levitte** [comment]
It's better, I'm getting a bit fewer of them.  #21 with current OpenSSL master gives me this:

[test-suite.log](https://github.com/tpm2-software/tpm2-openssl/files/6806975/test-suite.log)

**Comment by gotthardp** [comment]
This evening, sir, you did it! You did it! You did it!
You said that you would do it And indeed you did.

With a temporary fix to the CMP KUR key trouble I reached 100% tests passed. So, once it is decided how to address that issue this project should be able to pass all tests again!

I am really grateful for your help and fixes to the OpenSSL. Without that I would be still struggling somewhere deep in the code.

**Comment by levitte** [comment]
Absolutely!

``` console
$ tpm2_getcap properties-variable
TPM2_PT_PERSISTENT:
  ownerAuthSet:              0
  endorsementAuthSet:        0
  lockoutAuthSet:            1
  reserved1:                 0
  disableClear:              0
  inLockout:                 0
  tpmGeneratedEPS:           0
  reserved2:                 0
TPM2_PT_STARTUP_CLEAR:
  phEnable:                  1
  shEnable:                  1
  ehEnable:                  1
  phEnableNV:                1
  reserved1:                 0
  orderly:                   1
TPM2_PT_HR_NV_INDEX: 0x10
TPM2_PT_HR_LOADED: 0x0
TPM2_PT_HR_LOADED_AVAIL: 0x3
TPM2_PT_HR_ACTIVE: 0x0
TPM2_PT_HR_ACTIVE_AVAIL: 0x40
TPM2_PT_HR_TRANSIENT_AVAIL: 0x3
TPM2_PT_HR_PERSISTENT: 0x6
TPM2_PT_HR_PERSISTENT_AVAIL: 0x12
TPM2_PT_NV_COUNTERS: 0x2
TPM2_PT_NV_COUNTERS_AVAIL: 0x4
TPM2_PT_ALGORITHM_SET: 0x0
TPM2_PT_LOADED_CURVES: 0x2
TPM2_PT_LOCKOUT_COUNTER: 0x0
TPM2_PT_MAX_AUTH_FAIL: 0x20
TPM2_PT_LOCKOUT_INTERVAL: 0x1C20
TPM2_PT_LOCKOUT_RECOVERY: 0x15180
TPM2_PT_NV_WRITE_RECOVERY: 0x0
TPM2_PT_AUDIT_COUNTER_0: 0x0
TPM2_PT_AUDIT_COUNTER_1: 0x0
$ tpm2_getcap properties-fixed
TPM2_PT_FAMILY_INDICATOR:
  raw: 0x322E3000
  value: "2.0"
TPM2_PT_LEVEL:
  raw: 0
TPM2_PT_REVISION:
  raw: 0x8A
  value: 1.38
TPM2_PT_DAY_OF_YEAR:
  raw: 0x8
TPM2_PT_YEAR:
  raw: 0x7E2
TPM2_PT_MANUFACTURER:
  raw: 0x494E5443
  value: "INTC"
TPM2_PT_VENDOR_STRING_1:
  raw: 0x496E7465
  value: "Inte"
TPM2_PT_VENDOR_STRING_2:
  raw: 0x6C000000
  value: "l"
TPM2_PT_VENDOR_STRING_3:
  raw: 0x0
  value: ""
TPM2_PT_VENDOR_STRING_4:
  raw: 0x0
  value: ""
TPM2_PT_VENDOR_TPM_TYPE:
  raw: 0x0
TPM2_PT_FIRMWARE_VERSION_1:
  raw: 0x1930001
TPM2_PT_FIRMWARE_VERSION_2:
  raw: 0x0
TPM2_PT_INPUT_BUFFER:
  raw: 0x400
TPM2_PT_HR_TRANSIENT_MIN:
  raw: 0x3
TPM2_PT_HR_PERSISTENT_MIN:
  raw: 0x7
TPM2_PT_HR_LOADED_MIN:
  raw: 0x3
TPM2_PT_ACTIVE_SESSIONS_MAX:
  raw: 0x40
TPM2_PT_PCR_COUNT:
  raw: 0x18
TPM2_PT_PCR_SELECT_MIN:
  raw: 0x3
TPM2_PT_CONTEXT_GAP_MAX:
  raw: 0xFFFF
TPM2_PT_NV_COUNTERS_MAX:
  raw: 0x80
TPM2_PT_NV_INDEX_MAX:
  raw: 0x800
TPM2_PT_MEMORY:
  raw: 0x6
TPM2_PT_CLOCK_UPDATE:
  raw: 0x400000
TPM2_PT_CONTEXT_HASH:
  raw: 0xB
TPM2_PT_CONTEXT_SYM:
  raw: 0x6
TPM2_PT_CONTEXT_SYM_SIZE:
  raw: 0x100
TPM2_PT_ORDERLY_COUNT:
  raw: 0xFF
TPM2_PT_MAX_COMMAND_SIZE:
  raw: 0xF80
TPM2_PT_MAX_RESPONSE_SIZE:
  raw: 0xF80
TPM2_PT_MAX_DIGEST:
  raw: 0x20
TPM2_PT_MAX_OBJECT_CONTEXT:
  raw: 0x3A0
TPM2_PT_MAX_SESSION_CONTEXT:
  raw: 0xF8
TPM2_PT_PS_FAMILY_INDICATOR:
  raw: 0x1
TPM2_PT_PS_LEVEL:
  raw: 0x0
TPM2_PT_PS_REVISION:
  raw: 0x103
TPM2_PT_PS_DAY_OF_YEAR:
  raw: 0x0
TPM2_PT_PS_YEAR:
  raw: 0x0
TPM2_PT_SPLIT_MAX:
  raw: 0x80
TPM2_PT_TOTAL_COMMANDS:
  raw: 0x64
TPM2_PT_LIBRARY_COMMANDS:
  raw: 0x64
TPM2_PT_VENDOR_COMMANDS:
  raw: 0x0
TPM2_PT_NV_BUFFER_MAX:
  raw: 0x800
TPM2_PT_MODES:
  raw: 0x0
```

**Comment by gotthardp** [comment]
Hooray! All tests passed now, so I am closing this to indicate to users we are "good to go" again. The TPM memory errors will be handled as a part of #4. Thanks again.
